src/io/dataset.cpp

```diff
@@ -1211,40 +1225,35 @@ void Dataset::ConstructHistogramsMultiVal(const data_size_t* data_indices,
       (num_bin + kAlignedSize - 1) / kAlignedSize * kAlignedSize;
   int n_data_block = 1;
   int data_block_size = num_data;
-  Threading::BlockInfo<data_size_t>(share_state->num_threads, num_data, 1024,
-                                    &n_data_block, &data_block_size);
-  const size_t buf_size =
-      static_cast<size_t>(n_data_block - 1) * num_bin_aligned * 2;
-  if (share_state->hist_buf.size() < buf_size) {
-    share_state->hist_buf.resize(buf_size);
-  }
-  auto origin_hist_data = hist_data;
-  if (share_state->is_use_subcol) {
-    hist_data = share_state->TempBuf();
-  }
+  int min_block_size = std::min<int>(static_cast<int>(0.3f * num_bin /
+    multi_val_bin->num_element_per_row()) + 1, 1024);
```

Comment:
it seems `min_block_size` is not related ot `num_data`, maybe we can calculate it one time in `share_state`, and reuse it.